Abstract NUT-04 and NUT-05 Payment Methods (#258)
cashubtc/nuts: 04.md

Reviewer: Add used in NUT-23

---

> Minting tokens is a two-step process: requesting a mint quote and minting new tokens. This document describes the general flow that applies to all payment methods, with specifics for each officially specified payment method provided in dedicated NUTs.

Reviewer: ```suggestion
Minting tokens is a two-step process: requesting a mint quote and minting new tokens. This document describes the general flow that applies to all payment methods, with specifics for each supported payment method provided in dedicated NUTs.
```

## General Flow

The minting process follows these steps for all payment methods:

1. The wallet requests a mint quote for the `unit` to mint, specifying the payment `method`.
2. The mint responds with a quote that includes a `quote` id and a payment `request`.
3. The user pays the `request` using the specified payment method.
4. The wallet then requests minting of new tokens with the mint, including the `quote` id and new `outputs`.
5. The mint verifies payment and returns blind signatures.

## Common Request and Response Formats

### Mint Quote

To request a mint quote, the wallet of `Alice` makes a `POST /v1/mint/quote/{method}` request where `method` is the payment method requested (e.g., `bolt11`, `bolt12`, etc.).

```http
POST https://mint.host:3338/v1/mint/quote/{method}
```

Depending on the payment method, the request structure may vary, but all methods will include at minimum:

```json
{
  "unit": <str_enum[UNIT]>
  // Additional method-specific fields may be required
}
```

The mint `Bob` responds with a quote that includes some common fields for all methods:

```json
{
  "quote": <str>,
  "request": <str>,
  "unit":  <str_enum[UNIT]>,

>   "state": <str_enum[STATE]>,

Reviewer: State may not be used for all of them. For example in the current draft of bolt12 I've removed it in favor of having an `amount_paid` field and an `amount_minted` field. I recommend we remove the state here and then specify it in bolt11 (23).

>   "state": <str_enum[STATE]>,

Reviewer 2: I think bolt12 does not have this state
```suggestion
```

  // Additional method-specific fields will be included
}
```

Where `quote` is the quote ID, `request` is the payment request for the quote, and `unit` corresponds to the value provided in the request.

> `state` is an enum string field with possible values `"UNPAID"`, `"PAID"`, `"ISSUED"`:

Reviewer 2: This would have to go to the bolt11 spec in `23.md`

}
```

`MintMethodSetting` indicates supported `method` and `unit` pairs and additional settings of the mint. `disabled` indicates whether minting is disabled.

> `MintMethodSetting` is of the form:

Reviewer 2: This is also bolt11 specific, maybe we can mention here that there are settings for each method and their structure is defined in each spec file.

  "unit": <str>,
  "min_amount": <int|null>,
  "max_amount": <int|null>,
  "options": <Object|null>
}
```

> `min_amount` and `max_amount` indicate the minimum and maximum amount for an operation of this method-unit pair.

Reviewer: ```suggestion
`min_amount` and `max_amount` indicate the minimum and maximum amount for an operation of this method-unit pair. `options` are method-specific and can be defined in method-specific NUTs.
```

  "unit": <str>,
  "min_amount": <int|null>,
  "max_amount": <int|null>,

>   // Additional method-specific settings can be added here

Reviewer: ```suggestion
```

Since we have the options method specific settings should be added there not in the top level struct.

  "unit": <str>,
  "min_amount": <int|null>,
  "max_amount": <int|null>,

>   "description": <bool|null>,

Reviewer: ```suggestion
```

Description is bolt11 specific.

Reviewer 2: I would suggest to move the entire json to bol11 and mention here that settings like additional supported features and transactions limits can be included.

>   // Additional method-specific settings can be added here

Reviewer: ```suggestion
options: <Object>
```

Thinking about this a bit more, I think its better if we define the minimum common settings (method, unit, min, max) and then have a generic object that can be defined by the specific method nut. This way for each new method we add it to the nut04 list of settings and not create a nut23 setting (nut23 using the current suggested number for bolt11).
